File: my-folder/1544-count-good-nodes-in-binary-tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def goodNodes(self, root: TreeNode, mx=-inf) -> int:
        if not root:
            return 0
        left = self.goodNodes(root.left, max(mx, root.val))
        right = self.goodNodes(root.right, max(mx, root.val))

        return left + right + (mx <= root.val)





        # counter = [0]

        # def dfs(node, maxValue):
        #     if not node:
        #         return

        #     if node.val >= maxValue:
        #         counter[0] += 1
        #         maxValue = node.val

        #     dfs(node.left, maxValue)
        #     dfs(node.right, maxValue)

        # dfs(root, float('-inf'))
        # return counter[0]

        # def dfs(node, maxValue):
        #     if not node:
        #         return 0

        #     count = 0
        #     if node.val >= maxValue:
        #         count = 1

        #     left_count = dfs(node.left, max(maxValue, node.val))
        #     right_count = dfs(node.right, max(maxValue, node.val))

        #     return left_count + right_count + count

        # return dfs(root, float('-inf'))

File: my-folder/1544-count-good-nodes-in-binary-tree/solution.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var goodNodes = function(root) {
    let counter = 0;

    const dfs = (node, maxValue) => {
        if(!node) return;

        if(node.val >= maxValue) {
            counter++;
            maxValue = node.val;
        }

        dfs(node.left, maxValue);
        dfs(node.right, maxValue);
    }
    dfs(root, -Number.MAX_VALUE);
    return counter;
};
